core/src/test/scala/unit/kafka/log/OffsetMapTest.scala

```diff
@@ -28,54 +28,302 @@ class OffsetMapTest {
   @Test
   def testBasicValidation(): Unit = {
     validateMap(10)
+    validateMap(10, strategy = "header", headerKey = "sequence")
+    validateMap(10, strategy = "timestamp")
     validateMap(100)
+    validateMap(100, strategy = "header", headerKey = "sequence")
+    validateMap(100, strategy = "timestamp")
     validateMap(1000)
+    validateMap(1000, strategy = "header", headerKey = "sequence")
+    validateMap(1000, strategy = "timestamp")
     validateMap(5000)
+    validateMap(5000, strategy = "header", headerKey = "sequence")
+    validateMap(5000, strategy = "timestamp")
   }
   
   @Test
-  def testClear(): Unit = {
+  def testInit(): Unit = {
     val map = new SkimpyOffsetMap(4000)
-    for(i <- 0 until 10)
-      map.put(key(i), i)
-    for(i <- 0 until 10)
+    map.init()
+
+    for (i <- 0 until 10)
+      map.put(new FakeRecord(key(i), i))
+
+    for (i <- 0 until 10) {
       assertEquals(i.toLong, map.get(key(i)))
-    map.clear()
-    for(i <- 0 until 10)
+      assertEquals(-1L, map.getVersion(key(i)))
+      assertEquals(true, map.shouldRetainRecord(new FakeRecord(key(i), i)))
+      assertEquals(false, map.shouldRetainRecord(new FakeRecord(key(i), i-1)))
+    }
+
+    map.init(Defaults.CompactionStrategy)
+    for (i <- 0 until 10) {
       assertEquals(map.get(key(i)), -1L)
+      assertEquals(map.getVersion(key(i)), -1L)
+    }
   }
   
   @Test
   def testGetWhenFull(): Unit = {
     val map = new SkimpyOffsetMap(4096)
+    map.init(Defaults.CompactionStrategy)
     var i = 37L  //any value would do
     while (map.size < map.slots) {
-      map.put(key(i), i)
+      map.put(new FakeRecord(key(i), i))
       i = i + 1L
     }
+
     assertEquals(map.get(key(i)), -1L)
+    assertEquals(map.getVersion(key(i)), -1L)
     assertEquals(map.get(key(i-1L)), i-1L)
+    assertEquals(map.getVersion(key(i-1L)), -1L)
+    assertNotEquals(map.latestOffset, i)
+    assertEquals(true, map.shouldRetainRecord(new FakeRecord(key(i-1L), i-1L)))
+  }
+  
+  @Test
+  def testTimestampLogCompaction(): Unit = {
+    val map = new SkimpyOffsetMap(4096)
+    map.init("timestamp")
+
+    // put offset keys from 0 - 9 with corresponding offset 0-9 respectively
+    for (i <- 0 until 10)
+      map.put(new FakeRecord(key(i), i, fakeTimestamp = System.currentTimeMillis()))
+
+    // map should hold offset 0 for key 0
+    assertEquals(0, map.get(key(0)))
+    assertTrue(map.getVersion(key(0)) > 0)
+    // latest offset in map should be 9
+    assertEquals(9, map.latestOffset)
+    // size should 10
+    assertEquals(10, map.size)
+
+    // putting latest value for key=0 first followed by old value
+    val fr1 = new FakeRecord(key(0), 10, fakeTimestamp = System.currentTimeMillis())
+    map.put(fr1)
+    val fr2 = new FakeRecord(key(0), 11, fakeTimestamp = System.currentTimeMillis()-1000)
```

Comment:
This is fine... I was just thinking how we can make it flaky proof... for whatever reason, before line 104, GC kicks in for 2 seconds... then when you call System.currentTimeMillis - 1000 might return a time after 0 th record key... One option is not to use System.currentTimeMillis, but pin to a specific date time and use it to get future time / past time... Let me know if you need clarification on my thinking...